### ISSUE TITLE
Shorten Display of Birthdays in Calendar

### ISSUE THREAD

**Issue by deMattin** [owncloud/contacts#1027]
The birthdays shown in Calendar are very long and in some languages don't begin with the name. So depending on the displayable space, in month or week view (on smartphone) there's only shown "Geburtstag v...." in german for example and you have to go to detail view to see, who has a birthday.

What do you think about replacing the "Geburtstag von {name}" or "{name}'s Birthday" with the UTF8 code for a "birthday cake" (1F382) following by name and year?
I've seen tis in other apps, too and it's very intuitive, I think.

I tested the look by replacing the translation in the l10n files for my language (german) with the cake-code string: "&#x1f382; {name}" and it displays and looks very fine.
But this way it only works for displaying in owncloud web app.
The most issue is with the birthday in a caldav synced calendar where `{name}'s Birthday` or in german the `Geburtstag von {name}` is shown and often cut off disadvantageous in some display scenarios.

__So there is a be a better place to replace this in source for ALL languages with UTF8 birthday cake symbol as part of the entry.
But it only makes sense to investigate further, if core team likes this idea.__

Btw. This is an issue with contacts and not calendar, because contacts generates these calendar entries in owncloud!

**Comment by deMattin** [comment]
My solution, that should work for any language with no issues in log and working.
I need also to update language files because I'm no coder and so don't know, how to rewrite the code in config.php, not to search the language files for translation.
So without also editing language files you get log entries like "translation not found".

For coder from team this would be a 2 line edit in few seconds in config.php also with escaped UTF8 character, but there seems to be no interest and so I do it with my skills and post it here, if someone other likes to change this, too ... ;)

Add in ``apps/contacts/l10n/`` somewhere a new translation:
```
    "{name}" : "{name}",
```
in both language files for your language (Germany: `de.js` and `de.json` - I don't know where the german `de_DE` files are used for ... not for calendar)

in `apps/contacts/lib/contact.php`
change line 766 from
```
				App::$l10n->t('{name}\'s Birthday')
```
to
```
				App::$l10n->t('{name}')
```
and line 784 from
```
			$vEvent->{'SUMMARY'} = $title . ' (' . $date->format('Y') . ')';
```
to
```
			$vEvent->{'SUMMARY'} = 'ðŸŽ‚ ' . $title . ' (' . $date->format('Y') . ')';
```
where ðŸŽ you have to fill in the birthday cake utf8 symbol (in linux gedit by typing CTRL+SHIFT+u + the utf hex value "1f382")

That's it and working in sync and calendar app.

**Comment by pejakm** [comment]
Thanks, version with 🎂 icon looks best to me. I didn't even had to add anything to translation files.

**Comment by deMattin** [comment]
If you don't add anything to translation file, it works, but there probably will be errors in log file (something like "translation not found").
This is the only reason to modify l10n translation files.

If somebody can modify the function not to pass throgh translation, there's no modification of l10n files neccessary and also no log errors.
My php skills are very low and so I don't know, how to modify the call.

**Comment by pejakm** [comment]
Screw the log file. :) 

Here's how it looks like on my phone:
https://framapic.org/KCeGerr2CmRL/5MsDb2JS

**Comment by deMattin** [comment]
Yes, it looks really nice on smartphone, because there this "UTF8 cake" looks even more stylish as in browser (owncloud app).

If log files doesn't matter, no need of changing l10n files ... ;)

**Comment by deMattin** [comment]
OK, I did it __without the need of modifying language files__.
Some try and error, some research and now I understand, how this l10n works ... ;)

You only need to change/patch the following:

in `apps/contacts/lib/contact.php`
change line 766 from
```
				App::$l10n->t('{name}\'s Birthday')
```
to
```
				'{name}'
```
and line 784 from
```
			$vEvent->{'SUMMARY'} = $title . ' (' . $date->format('Y') . ')';
```
to
```
			$vEvent->{'SUMMARY'} = '🎂 ' . $title . ' (' . $date->format('Y') . ')';
```
where 🎂 you have to fill in the birthday cake utf8 symbol in editor (in linux gedit by typing CTRL+SHIFT+u + the utf hex value "1f382" - it may look weird there, but works)

So only one open question: How to define/escape this UTF8 cake sign in Hex-Code so that it would be possible to change this "hack" into an official pull request?

**Comment by gvde** [comment]
I doubt this is a good idea. It's a very small picture and may cause readability issues for some people.

For translations which currently don't have the name at the beginning it's a question of translation. For example, iOS uses "Firstname Lastname (x. Geburtstag)".

Or even simply use "Firstname Lastname (x)" without any need for translation. Using calendar colours the user should be able to identify which one-day events are birthdays and which are normal events...

**Comment by deMattin** [comment]
I think, most Calendars do it like that with a symbol.
And you don't need any text there - the calendar has it's own color.
So even without a symbol it is better than this: "Geburtstag von Walter Meier (1955)" like it is  actually in German.
This text is cropped in any Phone's calendar in overviews to "Geburtstag v....".

As an example, the birthday calender in Android generated from contacts only displays the name - even without any symbol. This is similar to it, you described for IOS - but without the "x. Birthday".
I think, you don't use the synced calendar of owncloud but the iOS functionality?

Readability? In Android it looks like @pejakm showed - very nice, I think.
And in Owncloud? Even a unreadyble dot would mark it enough and the UTF8 cake is "readable", if the text is readable. You don't need to identify every dot in a symbol. And it's an official UTF8 symbol designed by professionals!

Try it, you will like it ... ;)

Are you able to fix the source changes to not put the UTF8 code in there but instad an utf8_encode or something like that? I tried many things, but I don't get it.

I give you an impression, how it looks like in a month view of calendar (click to see it 1:1):
![image](https://cloud.githubusercontent.com/assets/5565503/9471638/979dbe42-4b51-11e5-836b-adee7640499d.png)

**Comment by gvde** [comment]
1. I don't think the birthday calendar in owncloud is supposed to be used with caldav. It's generated each time you access it. I think most clients today support contacts and calendaring and generate the birthday events in any way they want. Thus, I would say the birthday calendar in the contacts app is only to be used in the owncloud calendar app.

2. For presentation in the web browser, less is more. Omitting the symbol and the space after the symbol leaves you even more room, maybe avoiding that wrapped line for "Dagmar" on the 17th.

3. So I would say it's either a rearranged translation or a fully shorted "name (age)" only. Of course, as it's the web frontend it could become a configurable option to choose between the one or the other...

**Comment by deMattin** [comment]
1. Android doesn't give extra information like "x Birthday" or the year of Birthday. So owncloud gives more information and I would like to use it instead.

2. Now it's MUCH more with this "{names}'s Birthday" or in german "Geburtstag von {name}" and I aggree, it has to be shorten. I don't think, this symbol and space does extend it much.

3. My patch at line 766 is essential (and l10n independent), I think.
But to give the user a selectable "use or not to use symbol" switch will never be accepted by core team and there would be much code to implement this.

Maybe @jancborchardt from owncloud GUI team can say something to this modification?

**Comment by gvde** [comment]
Re 1. don't compare mobile client with web browser presentation. Both are designed with completely different goals.
Re 2. Translations can be changed. Why do you need a symbol where the colour can already says everything? It's not like those events are usually the most important in your calendar. Thus why do you need two markups for those? From the view of GUI design I don't think it's a good idea.
Re 3. The calendar and contacts apps are not part of the core anymore. The apps don't have a maintainer at the moment which is why there is a stall with these apps. IMHO it doesn't so much depend on what the core team will say but more whoever will take a lead on these apps and becomes maintainer...

**Comment by deMattin** [comment]
Yeah ok, you are right. Sadly, this app is without official maintainer and probably will never be any more.
It is missing some functionality but it works really well and cooperates with caldav and carddav syncing to phone without big issues.
And thanks to you, some of the remaining issues can be fixed or patched.

I tried "next gen apps" contactsplus and calendarplus but there are so many issues to be solved before it works like these official apps in basic functionality.
(and the birthday calendar there has a birthday cake in owncloud view)

It's no problem for me to modify any new (bugfix) release with my patch or hack (there are some others to do, too, I have to apply after every update) but all my cloud users liked it and so I thought, others would do, too.

So all in all you think, "no bug - no changes"?

You see presentation different than me. I like this symbol both in owncloud web and smartphone.

**Comment by jancborchardt** [comment]
Totally agree with @deMattin here that the display should simply be:
> :birthday: Firstname Lastname (19XX)

cc @libasys @georgehrke for Calendar.

**Comment by Tost69** [comment]
I'm not familiar with PHP but I've done it that way:

$bCake = (string)"\xf0\x9f\x8e\x82";
$vEvent->{'SUMMARY'} = $bCake . ' ' . $title . ' (' . $date->format('Y') . ')';

Cheers, Tom

**Comment by deMattin** [comment]
Thank you very much!
It's working well and now there's no need to hack UTF8 characters with editor any more.

So here's the complete and language independend patch to:
Shorten the display of birthdays in calendar
--
from **NAME's Birthday (YEAR)**
to **🎂 NAME (YEAR)**

Change/patch in `apps/contacts/lib/contact.php`:

replace **line 766**:
```
				App::$l10n->t('{name}\'s Birthday')
```
with:
```
				'{name}'
```
and
replace **line 784**:
```
			$vEvent->{'SUMMARY'} = $title . ' (' . $date->format('Y') . ')';
```
with:
```
			$bCake = (string)"\xf0\x9f\x8e\x82";
			$vEvent->{'SUMMARY'} = $bCake . ' ' . $title . ' (' . $date->format('Y') . ')';
```

**Comment by jancborchardt** [comment]
@deMattin please submit a pull request with that change. Thanks! :)

**Comment by deMattin** [comment]
Done ;)

**Comment by jancborchardt** [comment]
@deMattin awesome! :) Looking forward to more cool bugfixes and pull requests from you. :)

**Comment by ownhardy** [comment]
This "my contact's birthdays calendar" is the only way I can see those useful birthdays on my PC calendar. I use Thunderbird with the Lithning calender plugin, and this syncs via CalDav to Owncloud. Please also have this usecase in mind as well!

I know of no alternative way of watching a cloud-synced calendar on a PC, except for using Google Calendar or other proprietary software instead of Owncloud... Even using CardDav sync from Thunderbird, wouldn't be a substitute, because for me an email adress book -and this one only could be synced- is a very different thing to my full contacts adressbook with phone numbers, street adresses, and birthdays, ...

And in the german translation, the text is too long, now. In my calender view in Thunderbird the names are never even displayed for this reason, I can alwys only read "Birthday of". My vote would be for a text like

   Mike Miller (*2010)

or

   Mike Miller (5)

for this 5ths birthday in 2015.

The color should be enough to show it's a birthday. (You might also omitt the star if you prefer.)

(Just my single penny. Sad to hear that those apps have currently no maintainer. That might eventually drive me over to Google, which frightens me already. I really hope someone will step in here...)

**Comment by jancborchardt** [comment]
@ownhardy no worries, there’s work being done to move the CalDAV/CardDAV into core for the next ownCloud release, and to reboot the Calendar and Contacts apps. :)
